src/analyzer.fs

```diff
@@ -183,11 +183,16 @@ let markWrites topLevel = // calculates hasExternallyVisibleSideEffects, for inl
                     // functions are processed in order, so this is initialized before use
                     | Declaration.Func f -> f.hasExternallyVisibleSideEffects
                     | _ -> true
-                hasExternallyVisibleSideEffect <- hasExternallyVisibleSideEffect || hasSideEffect
+                if hasSideEffect then sideEffectIdents.Add(v) |> ignore<bool>
+                e
+            | FunCall(Var v, _) as e ->
+                if Builtin.pureBuiltinFunctions.Contains(v.Name) then
+                    // Visiting the Var previously found this ident to be a side effect, but now we know it's not.
```

Comment:
Why did we previously add it?
Can we check pureBuiltinFunctions when we see the ident?

Reply:
We added it when we saw the Var. We can't check pureBuiltinFunctions there because we don't know yet if it's a variable or a function. It's legal to write `float max = max(2.0, 3.0);`

Reply:
Yes, but we check the declaration above.

Besides, that would detect the case where the user defines a max function.